source/opt/insert_extract_elim.cpp

```diff
@@ -63,6 +77,124 @@ bool InsertExtractElimPass::IsVectorType(uint32_t typeId) {
+    if (cinst->opcode() == SpvOpCompositeInsert) {
+      if (ExtInsMatch(*pExtIndices, cinst, extOffset)) {
+        // Match! Use inserted value as replacement
+        replId = cinst->GetSingleWordInOperand(kInsertObjectIdInIdx);
+        break;
+      }
+      else if (ExtInsConflict(*pExtIndices, cinst, extOffset)) {
+        // If extract has fewer indices than the insert, stop searching.
+        // Otherwise increment offset of extract indices considered and
+        // continue searching through the inserted value
+        if (pExtIndices->size() - extOffset < cinst->NumInOperands() - 2) {
+          break;
+        } else {
+          extOffset += cinst->NumInOperands() - 2;
+          cid = cinst->GetSingleWordInOperand(kInsertObjectIdInIdx);
+        }
+      }
+      else {
+        // Consider next composite in insert chain
+        cid = cinst->GetSingleWordInOperand(kInsertCompositeIdInIdx);
+      }
+    } else if (cinst->opcode() == SpvOpVectorShuffle) {
+      // Get length of vector1
+      uint32_t v1_id =
+        cinst->GetSingleWordInOperand(kVectorShuffleVec1IdInIdx);
+      ir::Instruction* v1_inst = get_def_use_mgr()->GetDef(v1_id);
+      uint32_t v1_type_id = v1_inst->type_id();
+      ir::Instruction* v1_type_inst =
+        get_def_use_mgr()->GetDef(v1_type_id);
+      uint32_t v1_len =
+        v1_type_inst->GetSingleWordInOperand(kTypeVectorLengthInIdx);
+      // Get shuffle idx
+      uint32_t comp_idx = (*pExtIndices)[extOffset];
+      uint32_t shuffle_idx = cinst->GetSingleWordInOperand(
+        kVectorShuffleCompsInIdx + comp_idx);
+      // If undefined, give up
+      // TODO(greg-lunarg): Return OpUndef
+      if (shuffle_idx == 0xFFFFFFFF)
+        break;
+      if (shuffle_idx < v1_len) {
+        cid = v1_id;
+        (*pExtIndices)[extOffset] = shuffle_idx;
+      } else {
+        cid = cinst->GetSingleWordInOperand(kVectorShuffleVec2IdInIdx);
+        (*pExtIndices)[extOffset] = shuffle_idx - v1_len;
+      }
+    } else if (cinst->opcode() == SpvOpExtInst &&
+        cinst->GetSingleWordInOperand(kExtInstSetIdInIdx) ==
+          get_module()->GetExtInstImportId("GLSL.std.450") &&
+        cinst->GetSingleWordInOperand(kExtInstInstructionInIdx) ==
+          GLSLstd450FMix) {
+      // If mixing value component is 0 or 1 we just match with x or y.
+      // Otherwise give up.
+      uint32_t comp_idx = (*pExtIndices)[extOffset];
+      std::vector<uint32_t> aIndices = { comp_idx };
+      uint32_t a_id = cinst->GetSingleWordInOperand(kFMixAIdInIdx);
+      ir::Instruction* a_inst = get_def_use_mgr()->GetDef(a_id);
+      uint32_t a_comp_id = DoExtract(a_inst, &aIndices, 0);
+      if (a_comp_id == 0)
+        break;
+      ir::Instruction* a_comp_inst = get_def_use_mgr()->GetDef(a_comp_id);
+      if (a_comp_inst->opcode() != SpvOpConstant)
+        break;
+      uint32_t u = a_comp_inst->GetSingleWordInOperand(kConstantValueInIdx);
+      float* fp = reinterpret_cast<float*>(&u);
```

Comment:
The specification for FMix says that the operands are floating point type, but does not specify their width.  You might need to add a check that it is 32-bit.

Reply:
K.